x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformRetryableActions.java

```diff
@@ -0,0 +1,204 @@
+        this.threadPool = threadPool;
+        this.auditor = auditor;
+    }
+
+    /**
+     * @param transformId the id of the Transform, used for auditing and unique identification in combination with name
+     * @param name the name of this action, used for auditing and unique identification in combination with transformId
+     * @param action the function to register.  The retryable action will continuously call this action by passing it the provided listener.
+     *               The listener in the consumer must be notified when the task succeeds or fails via the {@link ActionListener#onResponse}
+     *               and {@link ActionListener#onFailure} API respectively.  If the listener is not called, the retry logic will break.
+     * @param listener this listener will be notified via #onResponse when the task succeeds or via #onFailure when there are no more
+     *                 retries left (or runWhile returns false).
+     * @param retryScheduledListener this listener will be notified after the first call. If true, another thread has started the retry
+     *                               process. If false, the original call was successful, and no retries will happen.
+     * @param runWhile this function will be called before every retry.  This can be used to halt the retry logic based on some boolean
+     *                 criteria.  When retries are stopped via this method, the retry logic will call the listener's onFailure method with
+     *                 the error causing the retries.
+     */
+    <Response> Runnable register(
+        String transformId,
+        String name,
+        Consumer<ActionListener<Response>> action,
+        ActionListener<Response> listener,
+        ActionListener<Boolean> retryScheduledListener,
+        Supplier<Boolean> runWhile
+    ) {
+        return unattendedTransformRetryables.compute(new TransformKey(transformId, name), (k, v) -> {
+            if (v != null) {
+                // do not create another retryable action for the same transform + function
+                logger.debug("[{}] {} already exists for transform", k.transformId(), k.functionName());
+                return v;
+            } else {
+                return new TransformRetryableAction<>(
+                    k.transformId(),
+                    action,
+                    ActionListener.runAfter(listener, () -> unattendedTransformRetryables.remove(k)),
+                    retryScheduledListener,
+                    runWhile
+                );
+            }
+        });
+    }
+
+    /**
+     * Stops all currently running retries and deregisters all retries for this transformId.
+     */
+    void stopAll(String transformId) {
+        // copy to an temp list, avoiding a modification exception since the cancel function will run a listener that removes from the map
+        retriesForTransform(transformId).toList().forEach(task -> {
```

Comment:
Yes, because `cancel` will eventually call the above listener that removes the entry from the map, so we'd be modifying the map during the iteration.  We need the extra List so that we can iterate over the tasks while updating the map.